### ISSUE TITLE
399. Evaluate Division Missing Optimal Union Find Solution

### ISSUE THREAD

**Issue by bschloss** [LeetCode-Feedback/LeetCode-Feedback#6497]
#### Your LeetCode username

bschloss

#### Category of the bug
- [ ] Question
- [x] Solution
- [ ] Language
- [ ] Missing Test Cases 

#### Description of the bug

The solution using union find for problem 399 does not use an optimal implementation of a disjoint set. Although paths are compressed, it does not make use of the Union by Rank technique. Consequently, the time complexity can be improved from (N + M) * O(Log(N)) to (N + M) * O(alpha(N)) ~ (N + M) * O(1). Also, since this is used in teaching for the disjoint set portion of the Graph explore card, having a solution which uses the optimal disjoint set implementation would be ideal. Unless, I have made a mistake, the problem essentially comes down to solving an algebra equation for rewriting the coefficients in terms of one root versus the other in the Union by Rank portion of the algorithm. When find(x) returns the root and the weight/coefficient relating the current node to the root, for the union algorithm we get

rootX, coef_x = self.find(x)
rootY, coef_y = self.find(y)

From the equations and values objects we know that:

x / y = value

so, if we rewrite x and y in terms of their roots, we have:

coef_x * rootX / (coef_y * rootY) = value

Consequently, if we wanted to rewrite rootX in terms of rootY or vice versa, we have:

rootX = (coef_y * value / coef_x) * rootY
rootY = (coef_x / value * coef_y) * rootX

Using these relationships, it is possible to add Union by Rank to the algorithm presented in the solution. I have provided my code that was submitted below. 

#### Code you used for Submit/Run operation

```
// Evaluate Division
import string
import collections


class DisjointSet:
    
    def __init__(self, nodes):
        self.node_to_number = {node: i for i, node in enumerate(nodes)}
        self.root = [i for i in range(len(nodes))]
        self.coef = [1] * len(nodes)
        self.rank = [1] * len(nodes)
    
    def find(self, x):
        if x == self.root[x]:
            return x, self.coef[x]
        self.root[x], coef = self.find(self.root[x])
        self.coef[x] *= coef 
        return self.root[x], self.coef[x]
    
    def union(self, node_x, node_y, value):
        x = self.node_to_number[node_x]
        y = self.node_to_number[node_y]
        rootX, coef_x = self.find(x)
        rootY, coef_y = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.root[rootY] = rootX
                self.coef[rootY] = coef_x / (value * coef_y)
            elif self.rank[rootX] < self.rank[rootY]:
                self.root[rootX] = rootY
                self.coef[rootX] = (coef_y * value) / coef_x
            else:
                self.root[rootY] = rootX
                self.coef[rootY] = coef_x / (value * coef_y)
                self.rank[rootX] += 1
    
    def evaluate(self, node_x, node_y):
        try:
            x = self.node_to_number[node_x]
            y = self.node_to_number[node_y]
            if x == y:
                return 1
            rootX, coef_x = self.find(x)
            rootY, coef_y = self.find(y)
            if rootX == rootY:
                return coef_x / coef_y
            else:
                return -1
        except KeyError:
            return -1


class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        nodes = set()
        for n1, n2 in equations:
            nodes.add(n1)
            nodes.add(n2)
        
        ds = DisjointSet(list(nodes))
        print(ds.node_to_number)
        for i in range(len(equations)):
            x, y = equations[i]
            value = values[i]
            ds.union(x, y, value)

        result = []
        for i in range(len(queries)):
            x, y = queries[i]
            result.append(ds.evaluate(x, y))
        return result
```

#### Language used for code

Python

#### Expected behavior

I expected to see a solution with both Union by Rank and path compression

#### Additional context

I have also commented on the solution with the code for others to see

**Comment by LC-Sue** [comment]
Hi @bschloss,
Thank you for sharing your insights. You are correct, we could improve the complexity by implementing union by rank with path compression. Although, I think the improvement might be less than you are expecting. The author here quotes the complexity of this union-find implementation as having an amortized time complexity for each operation of O(Log(N)) where Log is the iterated logarithm (not log):

https://en.wikipedia.org/wiki/Iterated_logarithm

Check out the Analysis of Algorithms section in the link above. It relates Log* to Log and the Inverse Ackermann Function.

TL;DR Log* and the Inverse Ackermann Function are quite similar in complexity.
